[quant][graph] Fix bug in replicateDequant

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37779 [quant][mobile] Return for conv with empty batch
* **#37637 [quant][graph] Fix bug in replicateDequant**
* #37635 [quant][graph] Fix bug in replaceConvolutionWithConv2d

Summary:
Insert dequant op at specific offset, rather than for all inputs of user

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21354931](https://our.internmc.facebook.com/intern/diff/D21354931)

supriyar committed:
[quant][graph] Fix bug in replicateDequant
Summary:
Insert dequant op at specific offset, rather than for all inputs of user

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Fix bug in replicateDequant"


Summary:
Insert dequant op at specific offset, rather than for all inputs of user

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1539,7 +1539,7 @@ void insertDeQuantForAllUse(
     WithInsertPoint ins(user);
     Node* dequant = insertDeQuant(graph, quantized_val, original_val, i);
-    user->replaceInputWith(original_val, dequant->output());
+    user->replaceInput(uses[i].offset, dequant->output());
```
Dont have the whole context here, but usually this means we missed a test that should have identified this issue. Is there a simple test that can be added?

jerryzh168 replied:
Yeah I think it's better to add a test.
this is pretty rare, it only happens when we do something like `torch.mul(x, x)`

supriyar committed:
Update on "[quant][graph] Fix bug in replicateDequant"


Summary:
Insert dequant op at specific offset, rather than for all inputs of user

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21354931](https://our.internmc.facebook.com/intern/diff/D21354931)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Fix bug in replicateDequant"


Summary:
Insert dequant op at specific offset, rather than for all inputs of user

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21354931](https://our.internmc.facebook.com/intern/diff/D21354931)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1199,6 +1199,23 @@ def forward(self, x, y, z):
+        qconfig_dict = {'': default_qconfig}
+        model = torch.jit.script(Mul()).eval()
+        m = quantize_script(model, qconfig_dict, _test_only_eval_fn, [data])
+        FileCheck().check_not("aten::mul") \
```
can you check for quantized::mul as well?

supriyar committed:
Update on "[quant][graph] Fix bug in replicateDequant"


Summary:
Insert dequant op at specific offset, rather than for all inputs of user

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21354931](https://our.internmc.facebook.com/intern/diff/D21354931)

[ghstack-poisoned]
